src/hotspot/share/opto/callGenerator.cpp

```diff
@@ -809,6 +813,424 @@ CallGenerator* CallGenerator::for_vector_reboxing_late_inline(ciMethod* method,
+              assert(addp1->is_AddP(), "");
+              assert(addp1->in(AddPNode::Base)->uncast()->is_Proj() &&
+                     addp1->in(AddPNode::Base)->uncast()->in(0)->as_CallJava()->method()->intrinsic_id() ==
+                     vmIntrinsics::_scopedValueCache, "");
+              if (scoped_value_cache == nullptr) {
+                scoped_value_cache = addp1->in(AddPNode::Base)->uncast()->in(0)->as_Call();
+              } else {
+                assert(scoped_value_cache == addp1->in(AddPNode::Base)->uncast()->in(0), "");
+              }
+              assert(in->in(MemNode::Memory)->is_Proj() && in->in(MemNode::Memory)->in(0) == scoped_value_cache, "");
+              Node* addp2 = addp1->in(AddPNode::Address);
+              Node* offset1 = addp1->in(AddPNode::Offset);
+              intptr_t const_offset = offset1->find_intptr_t_con(-1);
+              BasicType bt = TypeAryPtr::OOPS->array_element_basic_type();
+              int shift = exact_log2(type2aelembytes(bt));
+              int header = arrayOopDesc::base_offset_in_bytes(bt);
+              assert(const_offset >= header, "");
+              const_offset -= header;
+
+              Node* index = kit.gvn().intcon(const_offset >> shift);
+              if (addp2->is_AddP()) {
+                assert(!addp2->in(AddPNode::Address)->is_AddP() &&
+                       addp2->in(AddPNode::Base) == addp1->in(AddPNode::Base),
+                       "");
+                Node* offset2 = addp2->in(AddPNode::Offset);
+                assert(offset2->Opcode() == Op_LShiftX && offset2->in(2)->find_int_con(-1) == shift, "");
+                offset2 = offset2->in(1);
+#ifdef _LP64
+                assert(offset2->Opcode() == Op_ConvI2L, "");
+                offset2 = offset2->in(1);
+                if (offset2->Opcode() == Op_CastII && offset2->in(0)->is_Proj() &&
+                    offset2->in(0)->in(0) == get_cache_iff) {
+                  ShouldNotReachHere();
+                  offset2 = offset2->in(1);
+                }
+#endif
+                index = kit.gvn().transform(new AddINode(offset2, index));
+              }
+
+              if (get_first_iff == nullptr) {
+                get_first_iff = c->as_If();
+                first_index = index;
+              } else {
+                assert(get_second_iff == nullptr, "");
+                get_second_iff = c->as_If();
+                second_index = index;
+              }
+            }
+          } else if (c->is_RangeCheck()) {
+            // Kill the range checks as they are known to always succeed
+            kit.gvn().hash_delete(c);
+            c->set_req(1, kit.gvn().intcon(1));
+            C->record_for_igvn(c);
+          } else if (c->is_CallStaticJava()) {
+            assert(slow_call == nullptr, "");
+            slow_call = c->as_CallStaticJava();
+            assert(slow_call->method()->intrinsic_id() == vmIntrinsics::_SVslowGet, "");
```

Comment:
I would move the assert to before the assignment, but that is a matter of taste.